tracing/options.go

```diff
@@ -0,0 +1,115 @@
+// Copyright 2022 CloudWeGo Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package tracing
+
+import (
+	"context"
+
+	"github.com/cloudwego/kitex/pkg/rpcinfo"
+	"go.opentelemetry.io/otel"
+	"go.opentelemetry.io/otel/metric"
+	"go.opentelemetry.io/otel/metric/global"
+	"go.opentelemetry.io/otel/propagation"
+	"go.opentelemetry.io/otel/trace"
+)
+
+const (
+	instrumentationName = "github.com/kitex-contrib/obs-opentelemetry"
+)
+
+// Option opts for opentelemetry tracer provider
+type Option interface {
+	apply(cfg *config)
+}
+
+type option func(cfg *config)
+
+func (fn option) apply(cfg *config) {
+	fn(cfg)
+}
+
+type config struct {
+	tracer trace.Tracer
+	meter  metric.Meter
+
+	tracerProvider    trace.TracerProvider
+	meterProvider     metric.MeterProvider
+	textMapPropagator propagation.TextMapPropagator
+
+	spanNameFormatter func(ctx context.Context) string
+
+	withStackTrace        bool
+	recordSourceOperation bool
+}
+
+func newConfig(opts []Option) *config {
+	cfg := defaultConfig()
+
+	for _, opt := range opts {
+		opt.apply(cfg)
+	}
+
+	cfg.meter = cfg.meterProvider.Meter(
+		instrumentationName,
+		metric.WithInstrumentationVersion(SemVersion()),
+	)
+
+	cfg.tracer = cfg.tracerProvider.Tracer(
+		instrumentationName,
+		trace.WithInstrumentationVersion(SemVersion()),
+	)
+
+	return cfg
+}
+
+func defaultConfig() *config {
+	return &config{
+		tracerProvider:    otel.GetTracerProvider(),
+		meterProvider:     global.GetMeterProvider(),
+		textMapPropagator: otel.GetTextMapPropagator(),
+		spanNameFormatter: func(ctx context.Context) string {
+			endpoint := rpcinfo.GetRPCInfo(ctx).To()
+			return endpoint.Method()
```

Comment:
这里是否可以参考  https://github.com/kitex-contrib/tracer-opentracing/blob/82c95c1e7f1b2dc264f743dcd50a0dce50058e16/server.go#L72
```
return endpoint.ServiceName() + "::" + endpoint.Method()
```

Reply:
> 这里是否可以参考 https://github.com/kitex-contrib/tracer-opentracing/blob/82c95c1e7f1b2dc264f743dcd50a0dce50058e16/server.go#L72
> 
> ```
> return endpoint.ServiceName() + "::" + endpoint.Method()
> ```

I feel that we can consider setting the span.name by following the otel's specifications：`$package.$service/$method`

https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/trace/semantic_conventions/rpc.md#span-name

Reply:
It seems that the server mode cannot get pacakge from invocation 

because kitex init an empty invocation is instantiated with rpcinfo.NewServerInvocation()

![image](https://user-images.githubusercontent.com/19374680/159473993-94e25875-0c02-45dd-b933-7940779418c2.png)